benchmarks/halo_finder/ArborX_HaloFinder.hpp

```diff
@@ -0,0 +1,264 @@
+  {
+    return attach(intersects(Sphere{w._M_view(i), w._r}), (int)i);
+  }
+};
+} // namespace Traits
+
+namespace HaloFinder
+{
+template <typename MemorySpace>
+struct CCSCallback
+{
+  using tag = ArborX::Details::InlineCallbackTag;
+  Kokkos::View<int *, MemorySpace> stat_;
+
+  KOKKOS_INLINE_FUNCTION
+  int representative(int const i) const
+  {
+    // ##### ECL license (see LICENSE.ECL) #####
+    int curr = stat_(i);
+    if (curr != i)
+    {
+      int next, prev = i;
+      while (curr > (next = stat_(curr)))
+      {
+        stat_(prev) = next;
+        prev = curr;
+        curr = next;
+      }
+    }
+    return curr;
+  }
+
+  template <typename Query, typename Insert>
+  KOKKOS_FUNCTION void operator()(Query const &query, int j,
+                                  Insert const &) const
+  {
+    int const i = ArborX::getData(query);
+
+    // only process edge in one direction
+    if (i > j)
+    {
+      // initialize to the first neighbor that's smaller
+      if (Kokkos::atomic_compare_exchange(&stat_(i), i, j) == i)
+        return;
+
+      {
+        // ##### ECL license (see LICENSE.ECL) #####
+        int vstat = representative(i);
+        int ostat = representative(j);
+
+        bool repeat;
+        do
+        {
+          repeat = false;
+          if (vstat != ostat)
+          {
+            int ret;
+            if (vstat < ostat)
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(ostat), ostat,
+                                                         vstat)) != ostat)
+              {
+                ostat = ret;
+                repeat = true;
+              }
+            }
+            else
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(vstat), vstat,
+                                                         ostat)) != vstat)
+              {
+                vstat = ret;
+                repeat = true;
+              }
+            }
+          }
+        } while (repeat);
+      }
+    }
+  }
+};
+
+// FIXME: have trouble switching halo views to Properties...
+// error: could not convert template argument ‘& findHalos<Kokkos::Cuda,
+// Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>, Kokkos::CudaSpace,
+// Kokkos::CudaSpace>’ from ‘<unresolved overloaded function type>’ to ‘void
+// (*)(Kokkos::Cuda, const Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>&,
+// Kokkos::View<int*, Kokkos::CudaSpace>&, Kokkos::View<int*,
+// Kokkos::CudaSpace>&, float, int)’
+template <typename ExecutionSpace, typename Primitives, typename MemorySpace>
+void findHalos(ExecutionSpace exec_space, Primitives const &primitives,
+               Kokkos::View<int *, MemorySpace> &halos_indices,
+               Kokkos::View<int *, MemorySpace> &halos_offset,
+               float linking_length, int min_size = 2)
+{
+  Kokkos::Profiling::pushRegion("ArborX:HaloFinder");
+
+  using clock = std::chrono::high_resolution_clock;
+
+  clock::time_point start_total, start;
```

Comment:
Will need to remove the timers before the merge. They are just there for convenience right now.